### PR TITLE
refactor related venues + event fetching for performance/efficiency

### PR DESCRIPTION
**Pull request by 0xdevalias**
partially fixes https://github.com/sparkletown/internal-sparkle-issues/issues/699

**Note:** This is an alternative / additional implementation plan to https://github.com/sparkletown/sparkle/pull/1367

## TODO

- [x] Implement our context/provider pattern for `useRelatedVenues` ([ref](https://github.com/sparkletown/sparkle/pull/1368#discussion_r629135576))
- [x] Add custom code timing/performance traces to the new code (using `utils/performance`)
  - [x] `fetchSovereignVenue` in `useSovereignVenueId` (fixed in https://github.com/sparkletown/sparkle/pull/1368/commits/a3b4399ef2d47672bd1e36c00ecc3bee561967eb)
  - [x] `fetchRelatedVenues` in `RelatedVenuesProvider` (fixed in https://github.com/sparkletown/sparkle/pull/1368/commits/733cd534d1d4c8c39bdaa52135455fef42067bc3)
  - [x] `fetchAllVenueEvents` in `useVenueEvents` (fixed in https://github.com/sparkletown/sparkle/pull/1368/commits/778bcc9bacafed1d5d74f8a8615e0c01a7395538)
  - [x] Anything else?
- [x] Add code to fetch events for a given array of `venueIds` ~using our 'chunked array in query' pattern~ (can't use that pattern as they are sub-collections) (fixed in https://github.com/sparkletown/sparkle/pull/1368/commits/6ecb7a3126838a5dfed4390598d18c22c7d7b0ed)
- [x] Refactor the codebase to make use of `useRelatedVenues` rather than legacy `useLegacyConnectRelatedVenues` ([ref](https://github.com/sparkletown/sparkle/pull/1368#discussion_r629133383)) (fixed in https://github.com/sparkletown/sparkle/commit/07a9e0e7a64dc56936c88853fd02eb79077e38a0 + https://github.com/sparkletown/sparkle/commit/e905c452a719589a6779047deccdf8c30aa93519 + https://github.com/sparkletown/sparkle/commit/b061f26feec4fe48b393ac38d67d814f528d6990)
- [x] Remove `useLegacyConnectRelatedVenues` once nothing needs it anymore (removed in https://github.com/sparkletown/sparkle/pull/1368/commits/4fecc7e204ff30200009a0ee52f23f74843806bf)
- [x] Anything else?

## Potentially-OOS?

- [ ] Figure out a better way/location to use `RelatedVenuesProvider` so that it only renders once in a given tree (currently seems to happen twice because of its usage in `WithNavigationBar`)
- [ ] Refactor the codebase to remove any places that load all of the venues into Redux ([ref](https://github.com/sparkletown/sparkle/pull/1368#discussion_r631600484))
- [ ] Can we enable any firebase/etc options to enable local caching of queries/related data? ([ref](https://github.com/sparkletown/sparkle/pull/1368#issuecomment-841616702))
- [ ] Can we enable any firebase/etc options to enable multiple queries to be made over a single websocket connection/similar? ([ref](https://github.com/sparkletown/sparkle/pull/1368#issuecomment-841616180))
- [ ] Add basic engineering script for retrieving/displaying all related venues + how they relate?
  - This is likely going to be most useful if we need to add the `worldId` or similar as proposed in https://github.com/sparkletown/sparkle/pull/1367
- [ ] Can we write some code to estimate the size/complexity of data being stored in redux at the moment?
  - Non-blocking, more curiosity to try and measure improvements with this new code

## OOS

- Adding validators to `anyVenueWithIdConverter` ([ref](https://github.com/sparkletown/sparkle/pull/1368#discussion_r629128871))

### CODE REVIEW

File: src/api/venue.ts [hunk 1]
```diff
@@ -34,3 +44,71 @@ export const setVenueLiveStatus = async ({
     })
     .finally(onFinish);
 };
+
+// TODO: return an array of all venueId's we find on the full way up to the sovereignVenue
+export const fetchSovereignVenueId = async (
```
Comment (0xdevalias) [inline src/api/venue.ts:49]: > TODO: return an array of all venueId's we find on the full way up to the `sovereignVenue`

File: src/api/venue.ts [hunk 2]
```diff
@@ -34,3 +44,71 @@ export const setVenueLiveStatus = async ({
...
+
+  if (previouslyCheckedVenueIds.includes(venueId))
+    throw new Error(
+      `Circular reference detected. '${venueId}' has already been checked`
+    );
+
+  return fetchSovereignVenueId(venue.parentId, [
+    ...previouslyCheckedVenueIds,
+    venueId,
+  ]);
```
Comment (0xdevalias) [inline src/api/venue.ts:67]: [note] This should be all that's needed to resolve the potential 'infinite recursion' edgecase bug described in https://github.com/sparkletown/internal-sparkle-issues/issues/699

File: src/api/venue.ts [hunk 3]
```diff
@@ -34,3 +44,71 @@ export const setVenueLiveStatus = async ({
...
+};
+
+export const fetchVenue = async (
+  venueId: string
+): Promise<WithId<AnyVenue> | undefined> => {
+  const venueDoc = await getVenueRef(venueId).get();
+
+  // TODO: Use proper data validation + firestore model converters to set this type rather than just forcing it with 'as'
+  //  see .withConverter(soundConfigConverter) in fetchSoundConfigs in src/api/sounds.ts as an example
+  const venue: AnyVenue | undefined = venueDoc.data() as AnyVenue | undefined;
```
Comment (0xdevalias) [inline src/api/venue.ts:77]: > TODO: Use proper data validation + firestore model converters to set this type rather than just forcing it with 'as'
> see `.withConverter(soundConfigConverter)` in `fetchSoundConfigs` in `src/api/sounds.ts` as an example

File: src/api/venue.ts [hunk 4]
```diff
@@ -34,3 +44,71 @@ export const setVenueLiveStatus = async ({
...
+): Promise<WithId<AnyVenue>[]> => {
+  const childVenuesSnapshot = await getVenueCollectionRef()
+    .where("parentId", "==", venueId)
+    // .withConverter()
+    .get();
+
+  if (childVenuesSnapshot.empty) return [];
+
+  // TODO: Use proper data validation + firestore model converters to set this type rather than just forcing it with 'as'
+  //  see .withConverter(soundConfigConverter) in fetchSoundConfigs in src/api/sounds.ts as an example
```
Comment (0xdevalias) [inline src/api/venue.ts:95]: > TODO: Use proper data validation + firestore model converters to set this type rather than just forcing it with 'as'
> see `.withConverter(soundConfigConverter)` in `fetchSoundConfigs` in `src/api/sounds.ts` as an example

File: src/api/venue.ts [hunk 5]
```diff
@@ -34,3 +44,71 @@ export const setVenueLiveStatus = async ({
...
+  venueId: string
+): Promise<WithId<AnyVenue>[]> => {
+  // TODO: find the sovereignVenueId
+  // TODO: fetch the sovereignVenue (note if fetchSovereignVenueId returned the venue and not just the id, we wouldn't need this extra step)
+  // TODO: fetch the sovereignVenue's children
+  // TODO: recursively fetch their children, etc till we have all of them
+  // TODO: return the final array
+
+  return [];
+};
```
Comment (0xdevalias) [inline src/api/venue.ts:114]: > TODO: implement fetchRelatedVenues

Plan:

> TODO: find the `sovereignVenueId`
> TODO: fetch the `sovereignVenue` (note if `fetchSovereignVenueId` returned the venue and not just the `id`, we wouldn't need this extra step)
> TODO: fetch the `sovereignVenue`'s children
> TODO: recursively fetch their children, etc till we have all of them
> TODO: return the final array

Comment (0xdevalias) [reply src/api/venue.ts:114]: > TODO: return the final array

Do we want to do this as an array? Or do we want to maintain some concept of the hierarchy of the data?

If the latter, then we would probably want to use some form of `VenueNode` type concept similar to:

```typescript
export interface VenueNode {
  parentVenue?: WithId<AnyVenue>;
  selfVenue: WithId<AnyVenue>;
  childVenues: WithId<AnyVenue>[];
}
```

This would allow us to neatly contain all of the relevant structure/layout details in a probably efficient manner. The efficiency of this probably depends on the size of the venue objects.. I believe it would definitely be efficient if we were just storing their IDs rather than the whole objects?

Comment (0xdevalias) [reply src/api/venue.ts:114]: > TODO: recursively fetch their children, etc till we have all of them

For this, we can probably use our 'chunked array-in query' pattern to fetch multiple venues at the same time, rather than issuing a request for every child `venueId`.

Comment (0xdevalias) [reply src/api/venue.ts:114]: > TODO: find the `sovereignVenueId`
> TODO: fetch the `sovereignVenue` (note if `fetchSovereignVenueId` returned the venue and not just the id, we wouldn't need this extra step)

It would probably be better to refactor `fetchSovereignVenueId` to be `fetchSovereignVenue` (and then we can have a version of it that gets just the `id` if we want/need)

**Review by 0xdevalias — COMMENTED** [review]
Thoughts/plan/TODOs for implementing a newer more efficient version of `fetchRelatedVenues`

File: src/api/venue.ts (hunk 5 above)
Comment (mike-lvov) [reply src/api/venue.ts]: > Do we want to do this as an array? Or do we want to maintain some concept of the hierarchy of the data?
> 
> If the latter, then we would probably want to use some form of VenueNode type concept similar to:

Given the possibility of having grandchildren/grandgrandchildren, this concept seems to be impossible to implement?

Comment (0xdevalias) [reply src/api/venue.ts]: Which concept? An array, or a hierarchy?

Both would be able to support any level of nested data?

File: src/api/venue.ts (hunk 3 above)
Comment (0xdevalias) [reply src/api/venue.ts]: fixed in https://github.com/sparkletown/sparkle/pull/1368/commits/684ea850233c48e27270daf046fd029c74aa2f3a

File: src/api/venue.ts (hunk 4 above)
Comment (0xdevalias) [reply src/api/venue.ts]: fixed in https://github.com/sparkletown/sparkle/pull/1368/commits/684ea850233c48e27270daf046fd029c74aa2f3a

File: src/api/venue.ts (hunk 1 above)
Comment (0xdevalias) [reply src/api/venue.ts]: fixed by @mike-lvov in https://github.com/sparkletown/sparkle/pull/1368/commits/1b0b472e3ceeab3f355dbbf4cdf1aae6612b170a

File: src/api/venue.ts [hunk 6]
```diff
@@ -34,3 +47,144 @@ export const setVenueLiveStatus = async ({
...
+          .where("parentId", "in", venueIdsChunk)
+          .withConverter(anyVenueWithIdConverter)
+          .get();
+
+        return childVenuesSnapshot.docs
+          .filter((docSnapshot) => docSnapshot.exists)
+          .map((docSnapshot) => docSnapshot.data());
+      }
+    )
+  ).then((result) => result.flat());
```
Comment (0xdevalias) [inline src/api/venue.ts:121]: [note] Implementation of 'chunked in array query' pattern

File: src/api/venue.ts [hunk 7]
```diff
@@ -34,3 +47,144 @@ export const setVenueLiveStatus = async ({
...
+  },
+
+  fromFirestore: (
+    snapshot: firebase.firestore.QueryDocumentSnapshot
+  ): WithId<AnyVenue> => {
+    // @debt Properly check/validate this data rather than using 'as'
+    //   return withId(AnyVenueSchema.validateSync(snapshot.data(), snapshot.id);
+
+    return withId(snapshot.data() as AnyVenue, snapshot.id);
+  },
```
Comment (0xdevalias) [inline src/api/venue.ts:189]: [debt] Ideally we would implement these validators as part of the converter as well, but given the state of our Venue types/validations/etc at the moment, I think that would be more pain than we have time for in this PR. As is, this is probably no worse off than any of the rest of our codebase at the moment (which also doesn't validate data when reading back out of the database)

File: src/hooks/useRelatedVenues.ts [hunk 8]
```diff
@@ -57,14 +62,67 @@ const emptyArray: never[] = [];
...
+
+  const {
+    loading: isRelatedVenuesLoading,
+    error: relatedVenuesError,
+    value: relatedVenues,
+  } = useAsync(async () => {
+    if (!sovereignVenueId) return [];
+
+    return fetchRelatedVenues(sovereignVenueId);
+  }, [sovereignVenueId]);
```
Comment (0xdevalias) [inline src/hooks/useRelatedVenues.ts:102]: [note] Newly implemented `useRelatedVenues`, making use of `react-use`'s `useAsync` to dramatically simplify the boilerplate required for handling async functions, _including_ properly handling error state in a way that we can simply expose it to users in the frontend.

- https://github.com/streamich/react-use
  - https://github.com/streamich/react-use/blob/master/docs/useAsync.md
  - https://github.com/streamich/react-use/blob/master/docs/useAsyncFn.md
  - https://github.com/streamich/react-use/blob/master/docs/useAsyncRetry.md

File: src/hooks/useRelatedVenues.ts [hunk 9]
```diff
@@ -80,9 +138,12 @@ interface UseConnectRelatedVenuesReturn {
...
-export const useConnectRelatedVenues: ReactHook<
-  UseConnectRelatedVenuesProps,
-  UseConnectRelatedVenuesReturn
+/**
+ * @deprecated use useRelatedVenues instead
+ */
+export const useLegacyConnectRelatedVenues: ReactHook<
+  UseLegacyConnectRelatedVenuesProps,
+  UseLegacyConnectRelatedVenuesReturn
 > = ({ venueId, withEvents = false }) => {
```
Comment (0xdevalias) [inline src/hooks/useRelatedVenues.ts:147]: [TODO] Refactor the codebase to make use of `useRelatedVenues` rather than this legacy `useLegacyConnectRelatedVenues`. Be sure to look for anything that is indirectly consuming the state via `useSelectors` as well. Once there is nothing left that is relying on this, then we can delete it.

**Note:** For the locations that are using the `withEvents` version of this, we will need to implement a new api function/hook/similar for fetching them in an efficient way. This may be end up getting handled in https://github.com/sparkletown/sparkle/pull/1367

File: src/hooks/useRelatedVenues.ts [hunk 10]
```diff
@@ -57,14 +62,67 @@ const emptyArray: never[] = [];
...
+  sovereignVenueId?: string;
+  isSovereignVenueIdLoading: boolean;
+  sovereignVenueIdError?: string;
+
+  relatedVenues: WithId<AnyVenue>[];
+  isRelatedVenuesLoading: boolean;
+  relatedVenuesError?: Error;
+}
+
+export const useRelatedVenues: ReactHook<
```
Comment (0xdevalias) [inline src/hooks/useRelatedVenues.ts:82]: [TODO] Implement our context/provider pattern for `useRelatedVenues` so that the one set of data can be shared among all child components that need to access it, rather than re-running it each time the hook is used.

See notes in https://github.com/sparkletown/internal-sparkle-issues/issues/699#issuecomment-834003796 for more details on this pattern (referenced below):

> > This is where we can leverage the context/provider pattern i've been using recently, eg.
> > 
> > - custom loading screens in `env/kotr` (https://github.com/sparkletown/sparkle/pull/1240, note, this was our first usage of this pattern, so isn't as clean as it could be)
> > - custom sounds (https://github.com/sparkletown/sparkle/pull/1269)
> > - reactions performance refactor (https://github.com/sparkletown/sparkle/pull/1323, I believe this is the latest/cleanest example of using this pattern)
> > - etc
> > 
> > See usage (note that the custom loading screen code isn't merged back into `staging`/our main codebase yet; see https://github.com/sparkletown/internal-sparkle-issues/issues/587):
> > 
> > - https://github.com/sparkletown/sparkle/search?q=createContext
> > - https://github.com/sparkletown/sparkle/search?q=useContext
> > 
> > _Originally posted by @0xdevalias in https://github.com/sparkletown/sparkle/pull/1367#discussion_r627837529_
> 
> _Originally posted by @0xdevalias in https://github.com/sparkletown/internal-sparkle-issues/issues/699#issuecomment-834014760_

**Review by 0xdevalias — COMMENTED** [review]
Notes, TODO, etc.

File: src/api/venue.ts (hunk 5 above)
Comment (mike-lvov) [reply src/api/venue.ts]: > Which concept? An array, or a hierarchy of nodes?

I meant that if you would want to return children, grandchildren .... etc, it would not be possible because you don't know how many layers are there?

But if you call all of those `childVenues`, it's fine

File: src/hooks/useRelatedVenues.ts (hunk 9 above)
Comment (0xdevalias) [reply src/hooks/useRelatedVenues.ts]: fixed in https://github.com/sparkletown/sparkle/commit/07a9e0e7a64dc56936c88853fd02eb79077e38a0 + https://github.com/sparkletown/sparkle/commit/e905c452a719589a6779047deccdf8c30aa93519 + https://github.com/sparkletown/sparkle/commit/b061f26feec4fe48b393ac38d67d814f528d6990 (and the supporting commits that lead up to those)

File: src/hooks/useRoom.ts [hunk 11]
```diff
@@ -22,6 +22,7 @@ export const useRoom = ({ room, venueName }: UseRoomProps) => {
 
   const roomUrl = room.url;
 
+  // @debt This selector relies on all venues in firebase being loaded into memory.. not very efficient
   const venues = useSelector(orderedVenuesSelector);
```
Comment (0xdevalias) [inline src/hooks/useRoom.ts:26]: [note, ?OOS?] `useRoom` is used by:

- `src/components/molecules/EventDisplay/EventDisplay.tsx`
- `src/components/templates/PartyMap/components/RoomAttendance.tsx`
- `src/components/templates/PartyMap/components/Map/MapRoom.tsx`
- `src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx`

File: src/pages/Admin/Admin.tsx [hunk 12]
```diff
@@ -81,6 +81,7 @@ const VenueList: React.FC<VenueListProps> = ({
   selectedVenueId,
   roomIndex,
 }) => {
+  // @debt This selector relies on all venues in firebase being loaded into memory.. not very efficient
   const venues = useSelector(orderedVenuesSelector);
```
Comment (0xdevalias) [inline src/pages/Admin/Admin.tsx:85]: [note, ?OOS?] `VenueList` is rendered by `Admin` which is used in:

- `src/components/organisms/AppRouter/AppRouter.jsx`

File: src/pages/Admin/Admin_v2.tsx [hunk 13]
```diff
@@ -29,7 +29,9 @@ const Admin_v2: React.FC = () => {
   const { user } = useUser();
   useAdminVenues(user?.uid);
 
+  // @debt This selector relies on all venues in firebase being loaded into memory.. not very efficient
   const venues = useSelector(orderedVenuesSelector);
```
Comment (0xdevalias) [inline src/pages/Admin/Admin_v2.tsx:33]: `Admin_v2` is used in:

- `src/components/organisms/AppRouter/AppRouter.jsx`

File: src/pages/Admin/VenueList/VenueList.tsx [hunk 14]
```diff
@@ -13,7 +14,8 @@ const VenueList: React.FC<VenueListProps> = ({
   selectedVenueId,
   roomIndex,
 }) => {
-  const venues = useSelector((state) => state.firestore.ordered.venues);
+  // @debt This selector relies on all venues in firebase being loaded into memory.. not very efficient
+  const venues = useSelector(orderedVenuesSelector);
```
Comment (0xdevalias) [inline src/pages/Admin/VenueList/VenueList.tsx:18]: `src/pages/Admin/VenueList/VenueList.tsx` doesn't appear to be used anywhere anymore.

File: src/pages/VenuePage/VenuePage.tsx [hunk 15]
```diff
@@ -64,26 +64,38 @@ const VenuePage: React.FC = () => {
...
   const eventRequestStatus = useSelector(isCurrentEventRequestedSelector);
 
+  useConnectUserPurchaseHistory();
   const userPurchaseHistory = useSelector(userPurchaseHistorySelector);
   const userPurchaseHistoryRequestStatus = useSelector(
     isUserPurchaseHistoryRequestedSelector
   );
 
+  // @debt we REALLY shouldn't be loading all of the venues collection data like this, can we remove it?
+  useFirestoreConnect("venues");
```
Comment (0xdevalias) [inline src/pages/VenuePage/VenuePage.tsx:83]: [note] See notes about this in review comments for `src/utils/selectors.ts`

File: src/utils/selectors.ts [hunk 16]
```diff
@@ -74,11 +74,18 @@ export const worldUsersByIdSelector: SparkleSelector<
...
+ * @debt Refactor all places that rely on this, then remove it from the codebase
  */
 export const venuesSelector: SparkleSelector<Record<string, AnyVenue>> = (
   state
 ) => state.firestore.data.venues || {};
 
+/**
+ * @deprecated This selector requires all of the venues data in firebase to be loaded into memory. Find a different way.
+ * @debt Refactor all places that rely on this, then remove it from the codebase
+ */
```
Comment (0xdevalias) [inline src/utils/selectors.ts:88]: [note, ?OOS?] This appears to be used in the following locations across the codebase (which may not necessarily rely on this specific connection):

```
src/types/Firestore.ts
  FirestoreData
    venues
      src/utils/selectors.ts
        venuesSelector
          makeVenueSelector
            src/pages/Admin/Admin.tsx

  FirestoreOrdered
    venues
      src/utils/selectors.ts
        orderedVenuesSelector
          src/components/organisms/UserProfileModal/UserProfileModal.tsx
          src/components/templates/Playa/Playa.tsx
          src/hooks/useRoom.ts
          src/pages/Admin/Admin.tsx
          src/pages/Admin/Admin_v2.tsx
          src/pages/Admin/VenueList/VenueList.tsx
```

File: src/types/Firestore.ts [hunk 17]
```diff
@@ -69,6 +69,11 @@ export interface FirestoreData {
   worldUsers?: Record<string, User>;
   venueChatMessages?: Record<string, VenueChatMessage>;
   venueEvents?: Record<string, VenueEvent>;
+
+  /**
+   * @deprecated This state requires all of the venues data in firebase to be loaded into memory. Find a different way.
+   * @debt Refactor all places that rely on this, then remove it from the codebase
+   */
   venues?: Record<string, AnyVenue>;
```
Comment (0xdevalias) [inline src/types/Firestore.ts:77]: [note] See notes about this in review comments for `src/utils/selectors.ts`

File: src/types/Firestore.ts [hunk 18]
```diff
@@ -83,21 +88,21 @@ export interface FirestoreOrdered {
...
   posterVenues?: WithId<PosterPageVenue>[];
   worldUsers?: Array<WithId<User>>;
   venueChatMessages?: Array<WithId<VenueChatMessage>>;
   venueEvents?: Array<WithId<VenueEvent>>;
+
+  /**
+   * @deprecated This state requires all of the venues data in firebase to be loaded into memory. Find a different way.
+   * @debt Refactor all places that rely on this, then remove it from the codebase
+   */
   venues?: Array<WithId<AnyVenue>>;
```
Comment (0xdevalias) [inline src/types/Firestore.ts:107]: [note] See notes about this in review comments for `src/utils/selectors.ts`

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 19]
```diff
@@ -124,10 +124,12 @@ export const UserProfileModal: React.FC = () => {
 };
 
 const SuspectedLocation: React.FC<{ user: WithId<User> }> = ({ user }) => {
+  // @debt This will currently load all venues in firebase into memory.. not very efficient
   useFirestoreConnect("venues");
```
Comment (0xdevalias) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:128]: [note, ?OOS?] `SuspectedLocation` is rendered by `UserProfileModal` only when `ENABLE_SUSPECTED_LOCATION` is set to `true` (it's currently `false`). I believe this is a legacy burn feature.

`UserProfileModal` rendered in `src/pages/VenuePage/TemplateWrapper.tsx`

File: src/components/templates/Playa/Playa.tsx [hunk 20]
```diff
@@ -128,7 +127,12 @@ const isPlaced = (venue: AnyVenue) => {
 const minZoom = () => (window.innerWidth - 2 * PLAYA_MARGIN_X) / PLAYA_WIDTH;
 
 const Playa = () => {
+  // @debt This will currently load all venues in firebase into memory.. not very efficient
   useFirestoreConnect("venues");
+  const venues = useSelector(orderedVenuesSelector);
```
Comment (0xdevalias) [inline src/components/templates/Playa/Playa.tsx:132]: [note, ?OOS?] This is a legacy template that is scheduled to be removed, so shouldn't be an issue. We can probably even comment this out to be certain about it not getting run if we want.

**Review by 0xdevalias — COMMENTED** [review]
[note] Some notes/comments/etc.

**Comment by 0xdevalias** [comment]
Can we enable any firebase/etc options to enable multiple queries to be made over a single websocket connection/similar?

**tl;dr:** Firestore uses `WebChannels` from Google's closure library, with the 2 relevant settings seeming to be:

- https://firebase.google.com/docs/reference/js/firebase.firestore.Settings#optional-experimentalautodetectlongpolling
- https://firebase.google.com/docs/reference/js/firebase.firestore.Settings#optional-experimentalforcelongpolling

---

**DeepDive:**

- https://stackoverflow.com/questions/50594200/does-the-firestore-sdk-use-a-connected-socket-to-make-its-requests-or-individual/50594588#50594588
- https://github.com/firebase/firebase-js-sdk/search?q=websocket
- https://firebase.google.com/docs/reference/js/firebase.firestore.Settings#optional-experimentalautodetectlongpolling
  - > Configures the SDK's underlying transport (`WebChannel`) to automatically detect if long-polling should be used. This is very similar to `experimentalForceLongPolling`, but only uses long-polling if required.
    - https://github.com/firebase/firebase-js-sdk/search?q=webchannel
      - https://github.com/firebase/firebase-js-sdk/tree/master/packages/webchannel-wrapper
        - https://google.github.io/closure-library/api/goog.net.WebChannel.html
          - > A WebChannel represents a logical bi-directional channel over which the client communicates with a remote server that holds the other endpoint of the channel. A WebChannel is always created in the context of a shared WebChannelTransport instance. It is up to the underlying client-side and server-side implementations to decide how or when multiplexing is to be enabled.
          - https://github.com/google/closure-library/blob/master/closure/goog/labs/net/webchannel.js#L269-L276
            - > This is an experimental feature to use WHATWG Fetch/streams (when supported) for the backchannel.
      - https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/platform/browser/connection.ts#L28
- https://firebase.google.com/docs/reference/js/firebase.firestore.Settings#optional-experimentalforcelongpolling

**Comment by 0xdevalias** [comment]
Can we enable any firebase/etc options to enable local caching of queries/related data?

tl;dr: Yes. Firebase persistence and/or bundles.

--

**DeepDive:**

- https://firebase.google.com/docs/firestore/manage-data/enable-offline#web-v8
- https://firebase.google.com/docs/reference/js/firebase.firestore.Firestore#settings
  - https://firebase.google.com/docs/reference/js/firebase.firestore.Settings#optional-cachesizebytes
- https://firebase.google.com/docs/reference/js/firebase.firestore.Firestore#enablepersistence
  - https://firebase.google.com/docs/reference/js/firebase.firestore.PersistenceSettings
- https://firebase.google.com/docs/reference/js/firebase.firestore.Firestore#loadbundle
  - https://firebase.googleblog.com/2021/04/firestore-supports-data-bundles.html
  - https://firebase.google.com/docs/firestore/solutions/serve-bundles
- https://firebase.google.com/docs/reference/js/firebase.firestore.Firestore#namedquery
- https://firebase.google.com/docs/reference/js/firebase.firestore.Firestore#disablenetwork
- https://firebase.google.com/docs/reference/js/firebase.firestore.Firestore#enablenetwork
- https://firebase.googleblog.com/2018/09/multi-tab-offline-support-in-cloud.html

File: src/api/venue.ts [hunk 21]
```diff
@@ -34,3 +48,144 @@ export const setVenueLiveStatus = async ({
...
+      checkedVenueIds: previouslyCheckedVenueIds,
+    };
+
+  if (previouslyCheckedVenueIds.includes(venueId))
+    throw new Error(
+      `Circular reference detected. '${venueId}' has already been checked`
+    );
+
+  if (maxDepth && maxDepth <= 0)
+    throw new Error("Maximum depth reached before finding the sovereignVenue.");
```
Comment (mike-lvov) [inline src/api/venue.ts:84]: What's the usecase for this? I think we always want to reach the sovereign venue no matter how far in the hierarchy it is

File: src/components/molecules/NavBar/NavBar.tsx [hunk 22]
```diff
@@ -201,7 +201,7 @@ const NavBar: React.FC<NavBarPropsType> = ({
               >
                 {navbarTitle} <span className="schedule-text">Schedule</span>
               </div>
-              <VenuePartygoers />
+              <VenuePartygoers venueId={venueId} />
```
Comment (mike-lvov) [inline src/components/molecules/NavBar/NavBar.tsx:204]: 👍

File: src/components/molecules/NavBar/NavBar.tsx [hunk 23]
```diff
@@ -315,7 +315,10 @@ const NavBar: React.FC<NavBarPropsType> = ({
         }`}
         onClick={hideEventSchedule}
       >
-        <SchedulePageModal isVisible={isEventScheduleVisible} />
+        <SchedulePageModal
+          venueId={venueId}
```
Comment (mike-lvov) [inline src/components/molecules/NavBar/NavBar.tsx:319]: 🎉

File: src/components/molecules/VenuePartygoers/VenuePartygoers.tsx [hunk 24]
```diff
@@ -28,11 +30,11 @@ export const VenuePartygoers = () => {
...
   }, [
-    isCurrentVenueLoaded,
-    parentVenue,
-    currentVenue,
-    numberOfRecentWorldUsers,
+    isLoading,
     isRecentWorldUsersLoaded,
+    parentVenue?.attendeesTitle,
+    currentVenue?.attendeesTitle,
+    numberOfRecentWorldUsers,
```
Comment (mike-lvov) [inline src/components/molecules/VenuePartygoers/VenuePartygoers.tsx:37]: [OOS] I think it should only show sovereign venue name. It will reduce the amount of logic here

File: src/hooks/useRoom.ts (hunk 11 above)
Comment (mike-lvov) [reply src/hooks/useRoom.ts:26]: To bring some context. It's used to decide if the provided string is a venue or an external link

**Review by mike-lvov — APPROVED** [review]
Left some comment/explanations/questions, but overall looks as expected.

File: src/api/venue.ts (hunk 21 above)
Comment (0xdevalias) [reply src/api/venue.ts:84]: Best practice patterns for recursive functions. It's optional and not currently used anyway, but I expect we may want to implement a 'sane upper limit' at some stage (eg. venues can't be nested more than 50 levels deep or something that most normal users are never going to hit, but that protects us/our code), at which point this would be used for it.

I'd almost go as far as to say that we should define that default (in `src/settings`) now, and make it the default value of this prop/option here.

Having this limit in place would also make it impossible to find yourself in infinite recursive loops, as it would 'bail out' once that limit is hit.

In frontend code like this it's slightly less important, if this was running in a backend function it'd be super important to prevent excess computation time/etc (be it by accident, or by malicious intent).

File: src/hooks/useRoom.ts (hunk 11 above)
Comment (0xdevalias) [reply src/hooks/useRoom.ts:26]: OOS, will handle this in a followup PR.

File: src/pages/Admin/Admin.tsx (hunk 12 above)
Comment (0xdevalias) [reply src/pages/Admin/Admin.tsx:85]: OOS, will handle this in a followup PR.

File: src/pages/Admin/Admin_v2.tsx (hunk 13 above)
Comment (0xdevalias) [reply src/pages/Admin/Admin_v2.tsx:33]: OOS, will handle this in a followup PR.

File: src/pages/Admin/VenueList/VenueList.tsx (hunk 14 above)
Comment (0xdevalias) [reply src/pages/Admin/VenueList/VenueList.tsx:18]: OOS, will handle this in a followup PR.

File: src/pages/VenuePage/VenuePage.tsx (hunk 15 above)
Comment (0xdevalias) [reply src/pages/VenuePage/VenuePage.tsx:83]: OOS, will handle this in a followup PR.

File: src/utils/selectors.ts (hunk 16 above)
Comment (0xdevalias) [reply src/utils/selectors.ts:88]: OOS, will handle this in a followup PR.

File: src/types/Firestore.ts (hunk 17 above)
Comment (0xdevalias) [reply src/types/Firestore.ts:77]: OOS, will handle this in a followup PR.

File: src/types/Firestore.ts (hunk 18 above)
Comment (0xdevalias) [reply src/types/Firestore.ts:107]: OOS, will handle this in a followup PR.

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 19 above)
Comment (0xdevalias) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx:131]: OOS, will handle this in a followup PR.

File: src/components/templates/Playa/Playa.tsx (hunk 20 above)
Comment (0xdevalias) [reply src/components/templates/Playa/Playa.tsx:132]: OOS, will handle this in a followup PR.

File: src/api/venue.ts (hunk 21 above)
Comment (0xdevalias) [reply src/api/venue.ts:84]: OOS, will handle this in a followup PR.

File: src/components/molecules/VenuePartygoers/VenuePartygoers.tsx (hunk 24 above)
Comment (0xdevalias) [reply src/components/molecules/VenuePartygoers/VenuePartygoers.tsx:37]: Agree OOS, probably agree in general. Need to think through the logic/intent here a bit more before I can make that call for certain though.

File: src/hooks/useRoom.ts (hunk 11 above)
Comment (0xdevalias) [reply src/hooks/useRoom.ts:26]: > To bring some context. It's used to decide if the provided string is a venue or an external link

Surely we can figure that out in a simpler way.

**Review by 0xdevalias — COMMENTED** [review]
Have skimmed through the diff changes here, still happy with them, I say LGTM! (and then iterate on the remaining OOS things/minor polish in a followup PR)
